samtranslator/model/api/api_generator.py

```diff
@@ -204,20 +206,81 @@ def _construct_stage(self, deployment, swagger):
         stage.TracingEnabled = self.tracing_enabled
 
         if swagger is not None:
-            deployment.make_auto_deployable(stage, self.remove_extra_stage, swagger)
+            deployment.make_auto_deployable(stage, self.remove_extra_stage, swagger, self.domain)
 
         if self.tags is not None:
             stage.Tags = get_tag_list(self.tags)
 
         return stage
 
+    def _construct_api_domain(self, rest_api):
+        """
+        Constructs and returns the ApiGateway Domain and BasepathMapping
+        """
+        if self.domain is None or \
+           self.domain.get('DomainName') is None or \
+           self.domain.get('CertificateArn') is None:
+            return None, None
+
+        logical_id = logical_id_generator.LogicalIdGenerator("", self.domain).gen()
+
+        domain = ApiGatewayDomainName('ApiGatewayDomainName' + logical_id,
+                                      attributes=self.passthrough_resource_attributes)
+        domain.DomainName = self.domain.get('DomainName')
+        endpoint = self.domain.get('EndpointConfiguration')
+
+        if endpoint is None or endpoint not in ['EDGE', 'REGIONAL']:
+            endpoint = 'REGIONAL'
+
+        if endpoint == 'REGIONAL':
+            domain.RegionalCertificateArn = self.domain.get('CertificateArn')
+        else:
+            domain.CertificateArn = self.domain.get('CertificateArn')
+
+        domain.EndpointConfiguration = {"Types": [endpoint]}
+
+        # Create BasepathMappings
+        if self.domain.get('BasePath') and isinstance(self.domain.get('BasePath'), string_types):
+            basepaths = [self.domain.get('BasePath')]
+        elif self.domain.get('BasePath') and isinstance(self.domain.get('BasePath'), list):
+            basepaths = self.domain.get('BasePath')
+            # Cloudformation restricts you from adding '/' with other paths. '/' is also the default basepath value
+            if '/' in basepaths:
+                basepaths = None
```

Comment:
Again, it seems like we're making decisions on behalf of the customer here. I'm not sure I understand the implications of this well enough to say that we're making the right decision.

Reply:
If you don't specify a basepath, cfn defaults to '/'. If you specify basepaths like so - ['/', 'get', 'foo'] api gw errors out with the message - "Only one base path mapping is allowed if the base path is empty.". It only allows one domain to one api if there is an empty basepath. We can either throw an error here, or ignore other paths of '/' is specified. What is your recommendation?

Reply:
Again, I don't like us trying to auto-correct customer template errors. We should either throw an error or pass it through to API GW to throw an error, whichever makes more sense.

Reply:
I think I will change it to passing it to API GW to error out.